### PR TITLE
[Database] Migrate to Postgres **Draft PR**

### PR DESCRIPTION
**Pull request by Revolyssup**
**Description**
- Replaces SQLITE with Postgres
- Changes Meshery base image to alpine(CGO support no longer needed)

Untitled Diagram drawio

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.

### CODE REVIEW

**Comment by Revolyssup** [comment]
Most checks will fail for now

**Comment by Revolyssup** [comment]
@theBeginner86 and @gr455 , Can you guys checkout and test out this PR?

**Comment by leecalcote** [comment]
@Revolyssup has the "System Reset" user flow on the Settings page of Meshery UI been accounted for here, too?

**Review by theBeginner86 — COMMENTED** [review]
While running this, I was observed following problems:

1. `Invalid query` for both none and remote provider: Screenshot 2022-10-05 at 1 31 12 PM
2.  Local provider seeded content didn't save in DB.

Screenshot 2022-10-05 at 1 46 16 PM

3. Moreover, none of the MeshSync data was saved in the DB

Screenshot 2022-10-05 at 1 44 28 PM

**Comment by Revolyssup** [comment]
> While running this, I was faced following problems:
> 
> 1. `Invalid query` for both none and remote provider: Screenshot 2022-10-05 at 1 31 12 PM
> 2. Local provider seeded content didn't save in DB.
> 
> Screenshot 2022-10-05 at 1 46 16 PM
> 
> 3. Moreover, none of the MeshSync data was saved in the DB
> 
> Screenshot 2022-10-05 at 1 44 28 PM

For 2 & 3, You're looking in the database in the container right? And not the .sql file that we previously used?

**Comment by Revolyssup** [comment]
> @Revolyssup has the "System Reset" user flow on the Settings page of Meshery UI been accounted for here, too?

GORM provides a good abstraction. All that is taken care by GORM itself as we are not writing raw queries anywhere in Meshery(except for a few places, where I have made the appropriate changes)

**Comment by theBeginner86** [comment]
> > While running this, I was faced following problems:
> > 
> > 1. `Invalid query` for both none and remote provider: Screenshot 2022-10-05 at 1 31 12 PM
> > 2. Local provider seeded content didn't save in DB.
> > 
> > Screenshot 2022-10-05 at 1 46 16 PM
> > 
> > 3. Moreover, none of the MeshSync data was saved in the DB
> > 
> > Screenshot 2022-10-05 at 1 44 28 PM
> 
> For 2 & 3, You're looking in the database in the container right? And not the .sql file that we previously used?

yeah the one running in container.

**Comment by theBeginner86** [comment]
> > While running this, I observed following problems:
> > 
> > 1. `Invalid query` for both none and remote provider: Screenshot 2022-10-05 at 1 31 12 PM
> > 2. Local provider seeded content didn't save in DB.
> > 
> > Screenshot 2022-10-05 at 1 46 16 PM
> > 
> > 3. Moreover, none of the MeshSync data was saved in the DB
> > 
> > Screenshot 2022-10-05 at 1 44 28 PM
> 
> @theBeginner86 For the 1st and third issue, can you go to the meshsync package being used by Meshery and inside /pkg/model/model.go remove the gorm tags from ResourceSpec.Attribute and ResourceStatus.Attribute? This will fix the issue as postgres doesn't allow string matching on json.I dont know how sqlite was circumventing this.

Sure, doing it right now!

**Comment by CyrineG** [comment]
@Revolyssup @theBeginner86 I did get the `invalid query` too. To remove the gorm tags in the meshsync package being used by Meshery, do I need to redeploy meshsync from the source code (using `make run`)? Or does that package exist somewhere else?

**Comment by theBeginner86** [comment]
> @Revolyssup @theBeginner86 I did get the `invalid query` too. To remove the gorm tags in the meshsync package being used by Meshery, do I need to redeploy meshsync from the source code (using `make run`)? Or does that package exist somewhere else?

@Revolyssup  already added a fix for gorm tags in this PR https://github.com/meshery/meshsync/pull/158

@CyrineG, for testing you can check out the branch of that meshsync PR in your local meshsync, then update `go.mod` of this meshery branch to point to your local meshsync clone (that includes the code changes from that meshsync PR) by using the `replace` keyword. Then you just need to run `make run` again into meshery so that the structs are now referred from your local MeshSync rather than the latest released one.

**Comment by Revolyssup** [comment]
@leecalcote The golangci-lint issue will keep failing, this is the golangci incompatibility issue with go1.19. This is being tracked as an open issue there.

**Comment by gr455** [comment]
@Revolyssup This postgres instance does conflict with the postgres instance from meshery cloud on dev environments. The exposed port as well as the container name

```
docker: Error response from daemon: Conflict. The container name "/pg" is already in use by container "77678c050adaf9f9156001b3109902621ff8213d086023b7def682a5931fc001". You have to remove (or rename) that container to be able to reuse that name.
```

Can you change the container name and the default port so this doesn't happen?

**Comment by Revolyssup** [comment]
> @Revolyssup This postgres instance does conflict with the postgres instance from meshery cloud on dev environments. The exposed port as well as the container name
> 
> ```
> docker: Error response from daemon: Conflict. The container name "/pg" is already in use by container "77678c050adaf9f9156001b3109902621ff8213d086023b7def682a5931fc001". You have to remove (or rename) that container to be able to reuse that name.
> ```
> 
> Can you change the container name and the default port so this doesn't happen?

You can override this in the Makefile for dev environment.

File: .github/workflows/build-ui-and-server.yml [hunk 1]
```diff
@@ -231,6 +231,8 @@ jobs:
           PORT: 9081
           ADAPTER_URLS: "localhost:10000 localhost:10001 localhost:10002 localhost:10003 localhost:10004 localhost:10009 localhost:10007"
         run: |
+          docker run -d -p 5432:5432 -e POSTGRES_PASSWORD=meshery --name pg postgres
```
Comment (leecalcote) [inline .github/workflows/build-ui-and-server.yml:234]:
```suggestion
          docker run -d -p 5432:5432 -e POSTGRES_PASSWORD=meshery --name meshery-db postgres
```

File: install/deployment_yamls/k8s/meshery-deployment.yaml [hunk 2]
```diff
@@ -35,5 +37,12 @@ spec:
           ports:
             - containerPort: 8080
           resources: {}
+        - env:
+            - name: POSTGRES_PASSWORD
```
Comment (leecalcote) [inline install/deployment_yamls/k8s/meshery-deployment.yaml:41]:
```suggestion
            - name: DATABASE_PASSWORD
```

File: install/deployment_yamls/k8s/meshery-deployment.yaml [hunk 3]
```diff
@@ -35,5 +37,12 @@ spec:
...
+              value: meshery
+          name: pg
```
Comment (leecalcote) [inline install/deployment_yamls/k8s/meshery-deployment.yaml:43]:
```suggestion
          name: meshery-db 
```

File: install/docker/docker-compose.yaml [hunk 4]
```diff
@@ -15,75 +20,81 @@ services:
...
-    ports:
-      - "10010:10010"
-  meshery-cilium:
-    image: layer5/meshery-cilium:stable-latest
-    labels:
-      - "com.centurylinklabs.watchtower.enable=true"
+  meshery-db:
+    image:  postgres:10.22-alpine
+    environment:
+      - "POSTGRES_PASSWORD=meshery"
```
Comment (leecalcote) [inline install/docker/docker-compose.yaml:26]:
```suggestion
      - "DATABASE_PASSWORD=meshery"
```

File: install/kubernetes/helm/meshery/values.yaml [hunk 5]
```diff
@@ -7,15 +7,21 @@ restartPolicy: Always
 
 image:
   repository: layer5/meshery
-  tag: stable-latest
+  tag: edge-latest
   pullPolicy: Always
 
+db:
+  tag: postgres:10.22-alpine
+  container_name: pg
```
Comment (leecalcote) [inline install/kubernetes/helm/meshery/values.yaml:15]:
```suggestion
  container_name: meshery-db
```

File: install/kubernetes/helm/meshery/values.yaml [hunk 6]
```diff
@@ -7,15 +7,21 @@ restartPolicy: Always
...
+  pullPolicy: Always
 env:
   EVENT: mesheryLocal
   SAAS_BASE_URL: https://meshery.layer5.io
   PROVIDER_BASE_URLS: https://meshery.layer5.io
   ADAPTER_URLS: meshery-istio:10000 meshery-linkerd:10001 meshery-consul:10002 meshery-kuma:10007 meshery-osm:10009 meshery-nginx-sm:10010 meshery-nsm:10004 meshery-app-mesh:10005 meshery-traefik-mesh:10006 meshery-cilium:10012
 
+dbenv:
+  POSTGRES_PASSWORD: meshery
```
Comment (leecalcote) [inline install/kubernetes/helm/meshery/values.yaml:24]:
```suggestion
  DATABASE_PASSWORD: meshery
```

File: install/start_db.sh [hunk 7]
```diff
@@ -0,0 +1,8 @@
+	if [ ! "$(docker ps -q -f name=pg)" ]; then
+		if [ "$(docker ps -aq -f status=exited -f name=pg)" ]; then
+			# cleanup
+			docker rm pg
+		fi
+    # run your container
+    docker run -d -p 5432:5432 -e POSTGRES_PASSWORD=meshery --name pg postgres
```
Comment (leecalcote) [inline install/start_db.sh:7]:
```suggestion
    docker run -d -p 5432:5432 -e DATABASE_PASSWORD=meshery --name meshery-db postgres
```

File: install/start_db.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+	if [ ! "$(docker ps -q -f name=pg)" ]; then
+		if [ "$(docker ps -aq -f status=exited -f name=pg)" ]; then
+			# cleanup
+			docker rm pg
```
Comment (leecalcote) [inline install/start_db.sh:4]:
```suggestion
			docker rm meshery-db
```

File: install/start_db.sh [hunk 9]
```diff
@@ -0,0 +1,8 @@
+	if [ ! "$(docker ps -q -f name=pg)" ]; then
+		if [ "$(docker ps -aq -f status=exited -f name=pg)" ]; then
```
Comment (leecalcote) [inline install/start_db.sh:2]:
```suggestion
		if [ "$(docker ps -aq -f status=exited -f name=meshery-db)" ]; then
```

File: mesheryctl/internal/cli/root/system/meshery.yaml [hunk 10]
```diff
@@ -0,0 +1,89 @@
+version: "3"
+services:
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/meshery.yaml:2]: This seems to be missing `meshery-db`.

File: server/models/sql-utils.go [hunk 11]
```diff
@@ -59,18 +58,26 @@ func setNewDBInstance() {
...
+			Password: pass,
+			Host:     host,
+			Port:     port,
+			Logger:   log,
+		})
+		if err != nil {
+			log.Error(ErrConnectingDatabase(err))
+			log.Info(fmt.Sprintf("retrying after %d", 10))
+			time.Sleep(10 * time.Second)
+			continue
```
Comment (leecalcote) [inline server/models/sql-utils.go:74]:
```suggestion
		if err != nil {
			log.Error(ErrConnectingDatabase(err))
			log.Info(fmt.Sprintf("retrying after %d", 5))
			time.Sleep(5 * time.Second)
			continue
```

File: server/models/error.go [hunk 12]
```diff
@@ -117,6 +118,9 @@ var (
 	ErrContextAlreadyPersisted = errors.New(ErrContextAlreadyPersistedCode, errors.Alert, []string{"kubernetes context already persisted with provider"}, []string{"kubernetes context already persisted with provider"}, []string{}, []string{})
 )
 
+func ErrConnectingDatabase(err error) error {
+	return errors.New(ErrConnectingDatabaseCode, errors.Alert, []string{"Could not connect to the database"}, []string{"", err.Error()}, []string{""}, []string{"Make sure the configurations are correct"})
```
Comment (leecalcote) [inline server/models/error.go:122]: Please improve upon this.

File: server/cmd/main.go [hunk 13]
```diff
@@ -77,7 +77,10 @@ func main() {
 	viper.SetDefault("COMMITSHA", commitsha)
 	viper.SetDefault("RELEASE_CHANNEL", releasechannel)
 	viper.SetDefault("INSTANCE_ID", &instanceID)
-
+	viper.SetDefault("DBHOST", "localhost")
```
Comment (leecalcote) [inline server/cmd/main.go:80]:
```suggestion
	viper.SetDefault("DBHOST", "meshery-db")
```

File: install/deployment_yamls/k8s/meshery-deployment.yaml (hunk 2 above)
Comment (Revolyssup) [reply install/deployment_yamls/k8s/meshery-deployment.yaml:41]: This is a postgres specific environment variable. Postgres expects it to be `POSTGRES_PASSWORD`

File: .github/workflows/build-ui-and-server.yml (hunk 1 above)
Comment (Revolyssup) [reply .github/workflows/build-ui-and-server.yml:234]: Making this change

File: install/deployment_yamls/k8s/meshery-deployment.yaml (hunk 3 above)
Comment (Revolyssup) [reply install/deployment_yamls/k8s/meshery-deployment.yaml:43]: Making this change

File: install/docker/docker-compose.yaml (hunk 4 above)
Comment (Revolyssup) [reply install/docker/docker-compose.yaml:26]: This is a postgres specific environment variable. Postgres expects it to be POSTGRES_PASSWORD

File: install/kubernetes/helm/meshery/values.yaml (hunk 5 above)
Comment (Revolyssup) [reply install/kubernetes/helm/meshery/values.yaml:15]: Making this change

File: install/start_db.sh (hunk 7 above)
Comment (Revolyssup) [reply install/start_db.sh:7]: Making this change

File: install/start_db.sh (hunk 8 above)
Comment (Revolyssup) [reply install/start_db.sh:4]: Making this change

File: install/start_db.sh (hunk 9 above)
Comment (Revolyssup) [reply install/start_db.sh:2]: Making this change

File: mesheryctl/internal/cli/root/system/meshery.yaml (hunk 10 above)
Comment (Revolyssup) [reply mesheryctl/internal/cli/root/system/meshery.yaml:2]: Making this change

File: server/models/sql-utils.go (hunk 11 above)
Comment (Revolyssup) [reply server/models/sql-utils.go:74]: Makes sense. Making this change

File: server/models/error.go (hunk 12 above)
Comment (Revolyssup) [reply server/models/error.go:122]: Okay, enhancing the error.

File: server/cmd/main.go (hunk 13 above)
Comment (Revolyssup) [reply server/cmd/main.go:80]: By default it should be localhost. It is overriden has meshery-db only in docker-compose. For k8s where it is in same pod, we want it to communicate over localhost where it by passes the networking stack and directly communicates via socket. Same for dev environment.

**Pull request closed without merging**